File: depthFirstSearch.js
// Search through a tree using depth first search and return a list of node names

class Node {
  constructor(name) {
    this.name = name;
    this.children = [];
  }

  addChild(name) {
    this.children.push(new Node(name));
    return this;
  }

  // O(V + E) time | O(V) space

  depthFirstSearch(array) {
    let siblings = this.children;
    array.push(this.name);
    for (const sibling of siblings) {
      sibling.depthFirstSearch(array);
    }
    return array;
  }
}

File: Python/depthFirstSearch.py
class Node:
    def __init__(self, name):
        self.children = []
        self.name = name

    def addChild(self, name):
        self.children.append(Node(name))
        return self

# O(V + E) time | O(V) space    

    def depthFirstSearch(self, array):
    siblings = self.children 
    array.append(self.name)
    for sibling in siblings:
        sibling.depthFirstSearch(array)
    return array